blackbox/test_gcs.py

```diff
@@ -0,0 +1,169 @@
+    version=(4, 0, 0)
+)
+
+class FakeGCSServer:
+    URLS = {
+        'Linux-x86_64': 'https://github.com/fsouza/fake-gcs-server/releases/download/v1.47.8/fake-gcs-server_1.47.8_Linux_amd64.tar.gz',
+        'Darwin-x86_64': 'https://github.com/fsouza/fake-gcs-server/releases/download/v1.47.8/fake-gcs-server_1.47.8_Darwin_amd64.tar.gz',
+        'Darwin-arm64': 'https://github.com/fsouza/fake-gcs-server/releases/download/v1.47.8/fake-gcs-server_1.47.8_Darwin_arm64.tar.gz',
+    }
+
+    CACHE_ROOT = Path(os.environ.get('XDG_CACHE_HOME', os.path.join(os.path.expanduser('~'), '.cache')))
+    CACHE_DIR = CACHE_ROOT / 'crate-gcs-tests'
+
+    def __init__(self):
+        self.path = self._fetch_and_extract()
+        self.process = None
+
+    def _fetch_and_extract(self):
+        gcs_dir = FakeGCSServer.CACHE_DIR / 'fake-gcs'
+        gcs_tarball = gcs_dir / 'fake-gcs.tar.gz'
+        if not os.path.exists(gcs_dir):
+            os.makedirs(gcs_dir, exist_ok=True)
+            gcs_url = FakeGCSServer.URLS[f'{platform.system()}-{platform.machine()}']
+            urlretrieve(gcs_url, gcs_tarball)
+            with tarfile.open(gcs_tarball) as t:
+                t.extractall(gcs_tarball.parent)
+            os.remove(gcs_tarball)
+            gcs_dir.chmod(0o755)
+
+        return gcs_dir / 'fake-gcs-server'
+
+    def run(self):
+        cmd = [self.path, '-backend', 'memory', '-scheme', 'http', '-host', '127.0.0.1']
+        self.process = Popen(
+            cmd,
+            stdin=DEVNULL,
+            stdout=PIPE,
+            stderr=PIPE,
+            universal_newlines=True
+        )
+
+    def close(self):
+        if self.process:
+            self.process.terminate()
+            self.process.communicate(timeout=10)
+            self.process = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+
+class GCSSnapshotIntegrationTest(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        crate_node.start()
+
+    @classmethod
+    def tearDownClass(cls):
+        crate_node.stop()
+
+    def test_snapshot_roundtrip_to_gcs_and_back(self):
+        with FakeGCSServer() as gcs:
+            t = threading.Thread(target=gcs.run)
+            t.daemon = True
+            t.start()
+            wait_until(lambda: is_up('127.0.0.1', 4443))
+
+            client = storage.Client(
+                credentials=AnonymousCredentials(),
+                project="test-project",
+                client_options={"api_endpoint": "http://127.0.0.1:4443"},
+            )
+
+            client.bucket("test-bucket").create()
+
+            with connect(crate_node.http_url) as conn:
+                c = conn.cursor()
+                c.execute('DROP TABLE IF EXISTS t1')
+                c.execute('CREATE TABLE doc.t1 (x int) clustered into 1 shards with (number_of_replicas = 0)')
+                c.execute('INSERT INTO t1 SELECT generate_series(1,100000)')
```

Comment:
```suggestion
                c.execute('INSERT INTO t1 SELECT generate_series(1, 500)')
```

Or is there a reason to add more documents?

Reply:
The failure with the ordering of the blobs only occurred once I increased the data size, but it's fixed now. So i guess we can reduce it again.